### PR TITLE
Extended archshift's implementation of http:C service

### PR DESCRIPTION
**Pull request by Zaneo**
Changed the RequestType to lowercase, because DELETE is defined in winnt.h which is included in winsock2.h.

Changed the behavior of GetDownloadSizeState to return the current size of the data that has been copied out of the response_data buffer.

Changed the behavior of ReceiveData,

If the buffer is larger than the downloaded data size wait for more to be downloaded. (check downloaded size instead?)

If the data is larger than the buffer, fill the buffer, and increment an internal counter to advance through the response data.

Tested with Homebrew Browser, was able to download app, and then run them.

Addressed some comments from #738

### CODE REVIEW

File: src/core/hle/service/http/http_c.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
+        return;
+    }
+
+    auto context = map_it->second.get();
+
+    std::lock_guard<std::mutex> lock(map_it->second->mutex);
+    auto& data = context->response_data;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: This should be left explicitly as a vector reference, since it makes it more obvious what the immediate type is when reading. You also got rid of the `const` for no reason.

File: src/core/hle/service/http/http.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    int repeats = 0;
+    curl_multi_perform(manager, &still_running);
+
+    do {
+        // Number of file descriptors. If this is zero, nothing happened to
+        // the active connection.
+        int numfds;
+
+        // Set the timeout to 1sec.
+        if (curl_multi_wait(manager, NULL, 0, 1000, &numfds) != CURLM_OK)
```
Comment (lioncash) [inline src/core/hle/service/http/http.cpp]: nullptr

File: src/core/hle/service/http/http.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            break;
+
+        if (numfds == 0) {
+            // We allow one repeat try--after which, we start to sleep the
+            // thread until the timeout runs out.
+            if (repeats++ > 1)
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (lioncash) [inline src/core/hle/service/http/http.cpp]: lol

File: src/core/hle/service/http/http.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+/// Current state of the HTTP request (API-exposed).
+enum class RequestState : u32 {
+    NOT_STARTED = 1,
+    IN_PROGRESS = 5,
+    READY       = 8,
+    TIMEOUT     = 10,
+};
+
+/** A helper struct that contains all the required information for a HTTP request.
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: Opening comments shouldn't be on the `/**` line

File: src/core/hle/service/http/http.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+#pragma once
+
+#include <vector>
+#include <mutex>
+#include <memory>
+#include <thread>
+#include <string>
+#include <unordered_map>
+#include <atomic>
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: These need to be alphabetically sorted properly

File: src/core/hle/service/http/http.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+extern std::unordered_map<ContextHandle, std::unique_ptr<HttpContext>> context_map;
+extern ContextHandle next_handle;
+
+/** Makes an HTTP request using libcurl. Ideally called as part of a new thread.
+ * @param context A pointer to the HttpContext that will be used by this request. This pointer must remain valid.
+ */
+void MakeRequest(HttpContext* context);
+
+/** Appends a name-value pair to a curl_slist*
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: Opening description shouldn't be on the `/**` line.

Also this is missing documentation for name and value.

File: src/core/hle/service/http/http.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    long current_content_length;      //< The amount in bytes that has been read out of the http response_data
+
+    HttpContext();
+    ~HttpContext();
+};
+
+extern std::unordered_map<ContextHandle, std::unique_ptr<HttpContext>> context_map;
+extern ContextHandle next_handle;
+
+/** Makes an HTTP request using libcurl. Ideally called as part of a new thread.
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: Opening description shouldn't be on the `/**` line.

File: src/core/hle/service/http/http.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    std::mutex mutex;
+
+    //--- Ongoing request management
+    RequestState state;               //< API-exposed current state of the HTTP request
+    std::unique_ptr<std::thread> request_thread;
+    std::atomic<bool> cancel_request;              //< True if the request's thread should be canceled ASAP
+
+    //--- Request data
+    std::string url;                  //< URL to the target server.
+    RequestType request_type;             //< Type of request that will be performed
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: misaligned comment

File: src/core/hle/service/http/http.h [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    Post_Alt = 6,
+    Put_Alt  = 7
+};
+
+/// Current state of the HTTP request (API-exposed).
+enum class RequestState : u32 {
+    NOT_STARTED = 1,
+    IN_PROGRESS = 5,
+    READY       = 8,
+    TIMEOUT     = 10,
```
Comment (lioncash) [inline src/core/hle/service/http/http.h]: This should be lower-cased as well if RequestType is.

File: src/core/hle/service/http/http_c.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+    next_handle++;
+}
+
+static void CloseContext(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: Magic number

File: src/core/hle/service/http/http_c.cpp [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+static void CancelConnection(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+static void GetRequestState(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+static void GetDownloadSizeState(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: and again

File: src/core/hle/service/http/http_c.cpp [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+static void BeginRequest(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: and again

File: src/core/hle/service/http/http_c.cpp [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+static void ReceiveData(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+    u32 buf_size = cmd_buff[2];
+    u8* buffer = Memory::GetPointer(cmd_buff[4]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    auto context = map_it->second.get();
+
+    std::lock_guard<std::mutex> lock(map_it->second->mutex);
+    const std::vector<u8>& data = context->response_data;
+
+    // TODO: Check if this should use context->downloaded_size instead of data.size()
+    if (buf_size > data.size()) {
+        cmd_buff[1] = 0xD840A02B;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+    u32 hdr_name_len = cmd_buff[2];
+    u32 hdr_val_len  = cmd_buff[3];
+    char* hdr_name_buf = reinterpret_cast<char*>(Memory::GetPointer(cmd_buff[5]));
+    char* hdr_val_buf  = reinterpret_cast<char*>(Memory::GetPointer(cmd_buff[7]));
+
+    std::string hdr_name = std::string(hdr_name_buf, hdr_name_len);
+    std::string hdr_val  = std::string(hdr_val_buf, hdr_val_len);
+
+    if (hdr_name.empty()) {
+        cmd_buff[1] = 0xD8E0A002;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+static void GetResponseStatusCode(Service::Interface* self) {
+    u32* cmd_buff = Kernel::GetCommandBuffer();
+
+    ContextHandle handle = static_cast<ContextHandle>(cmd_buff[1]);
+
+    auto map_it = context_map.find(handle);
+    if (map_it == context_map.end()) {
+        cmd_buff[1] = 0xD8E007F7;
```
Comment (lioncash) [inline src/core/hle/service/http/http_c.cpp]: magic number

File: src/core/hle/service/http/http_c.cpp (hunk 17 above)
Comment (Zaneo) [inline src/core/hle/service/http/http_c.cpp]: @lioncash this seems to be the only one I can't find a definition for.

The others are RESULT_BADHANDLE, and RESULT_DOWNLOADPENDING.

File: src/core/hle/service/http/http.cpp (hunk 2 above)
Comment (Zaneo) [inline src/core/hle/service/http/http.cpp]: @lioncash numfds, it should be initialized first?

File: src/core/hle/service/http/http_c.cpp (hunk 17 above)
Comment (Subv) [inline src/core/hle/service/http/http_c.cpp]: Construct the ResultCode using ErrorDescription, ErrorLevel, etc

**Comment by bunnei** [comment]
@Zaneo do you plan on finishing this PR? Given that you have not touched it in months; I'm guessing not. Please let me know, otherwise this will be closed imminently.

**Comment by Zaneo** [comment]
Starting work on it again

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-13T11:04:23Z: https://github.com/citra-emu/citra/compare/ac0104b8b0584023cfbd7a4800f1ca1db890188f...Zaneo:72b4bc318c9f04cfa0f97590ac9db32349b72274

**Comment by neobrain** [comment]
Hi, this is neobot, using neobrain's account. I'm keeping an archive of versions of this PR:

2016-01-13T11:07:58Z: https://github.com/citra-emu/citra/compare/ac0104b8b0584023cfbd7a4800f1ca1db890188f...Zaneo:d3096e7a334940b0febdf1b81c1f0ab1c4cf9d1c

**Comment by Zaneo** [comment]
One of the issues, I haven't had a chance to look at is the libcurl.dll is not copied to the same location as the binary by default.

**Comment by LittleWhite-tb** [comment]
Check how this is done with Qt .DLL files (in CMakeLists.txt), you should have a clue on how to solve it.

File: .gitmodules [hunk 19]
```diff
@@ -7,3 +7,6 @@
 [submodule "nihstro"]
 	path = externals/nihstro
 	url = https://github.com/neobrain/nihstro.git
+[submodule "curl"]
+	path = externals/curl
+	url = https://github.com/bagder/curl.git
```
Comment (neobrain) [inline .gitmodules]: Is the version that we check out specified anywhere? If not, is there a way we could make sure only stable releases of curl are used, rather than bleeding-edge versions?

File: src/core/hle/result.h [hunk 20]
```diff
@@ -158,7 +160,7 @@ enum class ErrorSummary : u32 {
     Canceled = 9,
     StatusChanged = 10,
     Internal = 11,
-
+    HTTP_Unknown_16_DownloadPending = 16, // DownloadPending uses this, but the summary is not known
```
Comment (neobrain) [inline src/core/hle/result.h]: What is DownloadPending? (Clearly it's not the error code because how would an error code use anything?)

**Comment by neobrain** [comment]
How much thought was put into making sure this code is properly sandboxed? I.e.

* How likely is it that closing emulation during some network transaction will leak resources?
* How likely is it that closing emulation during some network transaction will in any way affect consecutive emulation sessions (without restarting Citra)? We wouldn't want network to fail when restarting a game just because network handles are still open or something, for instance.
* How likely is that a corrupted/malicious server response could introduce buggy behavior? It's probably overambitious to reach for a fully secure implementation in that aspect, but potential for "odd" network communication should still be considered in the design at least.
* How likely is it that the emulator will freeze indefinitely due to internal code waiting for some network connection? This would become important e.g. when the host network is in some way unstable (or if the user intentionally turns off internet access for whatever reason).

File: src/core/hle/result.h (hunk 20 above)
Comment (Zaneo) [inline src/core/hle/result.h]: The error code is not listed on http://3dbrew.org/wiki/Error_codes, and the only known ResultCode to use it is DownloadPending at this time. It could be left as, HTTP_Unknown_16, but I felt like this conveyed what was using it.

Comment (neobrain) [inline src/core/hle/result.h]: Let me rephrase: the comment doesn't make any sense. How can an error code (DownloadPending) "use" anything?

**Comment by Zaneo** [comment]
I clearly need to look into this a lot more.
